### ISSUE TITLE
The new runHttp2Client makes it hard to use a long-running http2 client

### ISSUE THREAD

**Issue by hce** [lucasdicioccio/http2-client#37]
I use http2-client in a library that needs to keep a long lived http2 connection open between multiple API calls. When using runHttp2Client to obtain a http2 connection, this isn't possible to achieve without a hack.

Could you offer an open and a close function in addition to the runHttp2Client function?

**Comment by lucasdicioccio** [comment]
I was wondering this design point for a while before making a new API using the bracketed pattern. Hence, thanks for your feedback.

One advantage of the bracketed pattern is to ensure that acquired resources are correctly cleaned and that no thread ever leaks (even if a server resets a connection super early in the setup process). The annoying bit of the bracketed pattern is that the "run function" must hijack the control of the thread to guarantee that a caller properly closes the function. So, overall there's some extra guarantee at the expense of a more annoying API.

I can provide both APIs but I'm afraid I won't be able to do so without some significant code duplication (and hence maintenance cost) given the current implementation. I'd recommend to stick to `0.3.0.2` for some time while I find time to improve the codebase to reduce the maintenance burden of having both APIs. I'm in lack of a good idea at the moment, hence I'll happily apply suggestions if you have some.

**Comment by hce** [comment]
To prevent threads and other resources from leaking I use the addFinalizer method of the System.Mem.Weak package. ( https://hackage.haskell.org/package/base-4.10.0.0/docs/System-Mem-Weak.html )

Couldn't you implement the bracketed version of the function by simply using and wrapping the non-bracketed functions, resulting in no code duplication? I didn't really understand why that wouldn't work in your case.

**Comment by lucasdicioccio** [comment]
The client is currently using a few Async per connection. As far as I can tell `withAsync` is the only and preferred way to properly clean-up when early asynchronous exception arises. I don't really enjoy the pattern of delegating the full-control of the current thread to the called library either, but so far I think it's the only one I want to support for properly handling something like "a server that immediately resets the connection". I can be wrong and maybe my code would correctly kill the background asyncs (e.g., because they wait on a no-longer-writeable Chan and get killed once the major GC discovers the deadlock) but I would not bet on it.

Do you mind trying breaking apart all the "async initialization steps" and provide an API that has the old-style of API? I'm happy to provide two APIs: one that "does the right thing" with an extra cost of delegating the control of the thread to the API, plus a "stochastically OK" API with an easier-to-use API and a disclaimer. Though, we should first try to extract out the initialization steps such as we don't duplicate the code (else it will be a challenge to keep the codebases in synch). I can give you a short guidance/walkthrough over Hangout if you want to give it a try.

**Comment by lucasdicioccio** [comment]
Hi, I've done the changes proposed in the 2nd paragraph of https://github.com/lucasdicioccio/http2-client/issues/37#issuecomment-341980364 ; 'newHttp2Client' should be back, and TLS is optional now. Please close this issue if you don't find regressions (I'm using 'runHttp2Client' myself and in the regression/example binary).

**Comment by hce** [comment]
Thanks a lot! Sorry for not replying earlier! I'll have a look at it right away and let you know if it works.

**Comment by hce** [comment]
Seems to be working fine! I'll run some longer lasting tests now, but in the meantime I'll close the ticket.

**Comment by lucasdicioccio** [comment]
Cool!

**Comment by hce** [comment]
Hmm, instead of a goaway notification, I seem to be getting this instead:

sendapn: <stdin>: hGetLine: end of file

I suspect that it comes from the http2-client library, but I am not sure. What do you think? Should I open a new ticket for that?

**Comment by lucasdicioccio** [comment]
A behavior change when a goaway is closely followed by a close is totally possible because the main loop will die before the interested threads get a chance to receive the notification. That said, the same behavior should have already been present -- maybe less likely -- in the previous version.

I'll try to reproduce the issue, even if it's in a lib underneath, the hGetLine should not leak as-is to users. Hence, please open a new bug, if possible with a `-xc` stack to have a better understanding of the context of the error: https://wiki.haskell.org/Debugging . Also, can you try against `master` (unreleased `0.5.0.1` -- I've changed to whole dispatching logic to avoid channels, the code is a bit more complex but we save a lot of RAM)? Then we can discuss how to make debugging simpler by printing sent/received frames much like I do in the example binary.

**Comment by lucasdicioccio** [comment]
Ok, I could get to a similar issue (e.g., on a non-http2 server). The problem definitely occurs when the TCP connection is reset before receiving a GoAway. Then, I cannot tell if you experience such an issue in your setup. Another situation where an equivalent error can happen is when the client closes the connection and the data-input-loop thread dies too.

Not sure really what to do because of the many way such an error could make the internal state go wonky. I've pushed a branch `handle-errors` as an intermediate patch to allow catching some of these exceptions but there could still be other reasons for exceptions. Also, the handling mechanism for such errors still is "interesting" because of the asynchronous-yet-serialized nature of HTTP2; rethrow-ing is not better, ignoring is not great either, etc.

**Comment by hce** [comment]
Thanks! I think for me the best solution is to catch all exceptions and invalidate the connection from the connection pool if an exception is caught.

**Comment by lucasdicioccio** [comment]
I've opened #49 to solve the problem in depth. Keep in mind that there are two ways to get unrecoverable exception:
- while reading on the socket (on the dispatcher thread)
- while writing on the socket (from all streams)
- goaways
Also, another type of exception that are a bit more recoverable (as in: we can still do _stuff_) are writing to half-closed/closed streams.
Finally, the potential decoding errors are interesting. Currently I'm exposing too much of these errors to library users and I may want to clean this up a bit.

I'll spend some time poking at various designs, trying to unify all these errors.

**Comment by hce** [comment]
Thanks!
Yeah, I was mistaken. Because the exception is thrown into the user's thread that called the function. So unless I use forkIO (which I'd rather not) there is no way for me to catch the exception.
